source/slang/slang.cpp

```diff
@@ -2689,6 +2693,16 @@ RefPtr<Module> Linkage::findOrImportModule(
         return loadedModule;
     }
 
+    // If the user is providing an additional list of loaded modules, we find
+    // if the module being imported is in that list. This allows a translation
+    // unit to use previously checked translation units in the same
+    // FrontEndCompileRequest.
+    Module* previouslyLoadedModule = nullptr;
+    if (loadedModules && loadedModules->TryGetValue(name, previouslyLoadedModule))
+    {
+        return previouslyLoadedModule;
+    }
```

Comment:
The one thing this doesn't handle is the case where the user specifies something like a single compile with:

* Translation unit 0 is for a module named `A` and includes the code `import B`
* Translation unit 1 is for a module named `B`

Because we process the translation units in the order they are specified, the `import B` in translation unit 0 will not be able to see module `B` because it hasn't been compiled/loaded yet. The user might be a bit surprised by that fact.

This implementation is fine for unblocking the user who requested this feature, but we should probably put some kind of `TODO` comment in here noting that a more complete long-term solution should detect cases where an earlier module in a compile is trying to import a later one from the same compile and either:

* Issue an error, indicating that the requested module was found, but isn't loaded yet because the compiler respects the user-specified order

* Allow such cases, by demand-loading the later translation unit and doing a recursive compile (just like we do for the more ordinary `import` case), and then skipping that translation unit when the compiler gets to it in the ordinary loop

The second option is implementable in principle, so it is *probably* the one we'd want to do in the long run, but it sounds kind of annoying.

The first option could actually be added to this solution pretty easily by installing a `nullptr` in `loadedModules` for all the module names specified in a compilation, and then replacing those entries as they get loaded. The logic here could then detect the null and issue an error.